cli: add flash-fast command

mndza opened the pull request:
This PR adds a command to Apollo CLI (`apollo flash-fast`) for flash programming. This command first programs a  gateware SPI bridge to the FPGA and then uses it to access the configuration SPI flash.

mndza commented:
True, let's wait until that other PR is ready (I'm marking as draft for avoiding confusion again).

I would like to create a PR with the `ApolloAdvertiser` injection code first (should that go in the `cynthion` repo?). That's why https://github.com/greatscottgadgets/luna/pull/220 is still a draft (it creates a manual instance of the advertiser).

I do not have a strong opinion on the location of the flash bridge gateware. If we decide the code belongs here, I'm willing to move it.

martinling commented:
There's three parts as I see it:
- support for platform-specific hooks, which needs to be added to `USBDevice` in the LUNA repo.
- the hooks, which need to be added to the platform definitions in the Cynthion repo.
- the `ApolloAdvertiser` gateware, which I think should live in `apollo_fpga.gateware` and can be imported from there by all Apollo-based platforms that need it (which will include Cynthion, URTI and others).

martinling commented on apollo_fpga/commands/cli.py:
```diff
@@ -16,12 +16,21 @@
 import logging
 import argparse
 from collections import namedtuple
+import xdg.BaseDirectory
```
This adds a dependency on `pyxdg` that isn't declared in `pyproject.toml`.

martinling commented on apollo_fpga/commands/cli.py:
```diff
@@ -16,12 +16,21 @@
+try:
+    from luna.gateware.platform import get_appropriate_platform
+    from apollo_fpga.gateware.flash_bridge import FlashBridge, FlashBridgeConnection
+except ImportError:
+    pass
```
This will stop the import failing, but if the user runs `apollo flash-fast` they'll just get a `NameError` later, right?

We should set a flag here, and if they try to use the feature, spot that flag and print a helpful message.

mndza commented:
Rebased.
- Changed `bInterfaceSubclass` to `0x01` to avoid conflict with Apollo stub interface.
- Added a final trigger reconfiguration step when closing the connection with the flash bridge. This makes `flash-fast` inconsistent with the `flash` command (which does not trigger a reconfiguration), but avoids keeping the Flash bridge enumerated.

mndza commented on apollo_fpga/gateware/flash_bridge.py:
```diff
@@ -0,0 +1,347 @@
+from usb_protocol.emitters            import DeviceDescriptorCollection
+
+VENDOR_ID  = 0x1d50
+PRODUCT_ID = 0x615b
```
Changed to pid.codes VID & test PID 000F

martinling reviewed (approved):
All good!
